src/mpi/request/waitall.c

```diff
@@ -88,14 +88,8 @@ int MPIR_Waitall_impl(int count, MPIR_Request * request_ptrs[], MPI_Status array
             }
             /* wait for ith request to complete */
             while (!MPIR_Request_is_complete(request_ptrs[i])) {
-                if (request_ptrs[i]->kind == MPIR_REQUEST_KIND__GREQUEST &&
-                    request_ptrs[i]->u.ureq.greq_fns->wait_fn != NULL) {
```

Comment:
We used to check if there was a `wait_fn` before calling it here. Now we are relying on the fact there is a `poll_fn` in `MPIX_Request_is_ex_generalized`. This might cause problems?

Reply:
I rely on the fact that the generalized requests are initialized with both callback; so checking one of which will be sufficient. An alternative is to make two different calls: `MPIX_Request_has_wait_fn` and `MPIX_Request_has_poll_fn`, what do you prefer ?

Reply:
I guess the question is, does MPICH require both functions to be supplied at creation time? From what I can see, we do not, so I think we need to check for a non-NULL `wait_fn` before calling it. I think the 2 different calls option is the best we can do for now.